### PR TITLE
refactor(schemas,core,cli): init tenant organizations

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

This is the beginning of the tenant organization series.

The main goal of this series is to leverage Logto Organizations for managing user tenants, replacing the current API resource-based approach. This will allow us to provide a better user experience for Logto Cloud, and it is also the foundation for collaboration features.

### Terms

- Tenant organization: An organization that represents a user tenant in the admin tenant.
- Management API proxy: An endpoint in Logto Cloud that proxies all requests to the corresponding tenant's Management API.

### Outline

- Database alteration script
- Common types, enums, and utils for tenant organizations
- CLI now supports only seeding test data for legacy versions

### How it works

When creating a new tenant (e.g. `foo`), the following resources will be created in the admin tenant:

```mermaid
graph LR
  subgraph Admin tenant
    subgraph Organizations
      foo("foo (i.e. tenant organization)")
    end
    subgraph API resources
      subgraph "http://foo.logto.app/api"
        scope(scope `all`)
      end
    end
    subgraph Applications
      app(Machine-to-machine app for tenant `foo`)
    end
    subgraph Roles
      role(Machine-to-machine role for tenant `foo`)
    end
  end

  app -.- role
  role -.- scope
```

Unlike the original RBAC approach, the user will now join the tenant organization `foo` for accessing the Management API proxy.

> [!Note]
> We still use the admin tenant issued access token to access user tenant. The subject changed from user to machine-to-machine app in the admin tenant.

The following diagram shows how Management API proxy works:

```mermaid
sequenceDiagram
  participant console as Logto Console
  participant cloud as Management API proxy (Logto Cloud)
  participant tenant as User tenant (Management API)
  participant admin as Admin tenant (Management API)

  console->>cloud: Request to get applications (GET /applications) for tenant `foo`
  break when user has not effective tenant organization permissions
    cloud->>console: Access denied
  end
  alt access token has expired
    cloud->>cloud: Get the machine-to-machine application for tenant `foo`
    cloud->>admin: Request access token for tenant `foo`
    admin->>cloud: Return access token
  end
  cloud->>tenant: Request `GET /applications` with access token
  tenant->>cloud: Return response
  cloud->>console: Return response
```

### Plan

The migration can have no downtime, with the following steps:

1. Temporarily disable registration for Logto Cloud. (manual)
2. Deploy database changes for tenant organizations.
3. Deploy the core and cloud services with the new changes.
4. Clean up the outdated user roles for tenant management.
5. Enable registration for Logto Cloud. (manual)

This pull request contains the changes for step 2.

## Testing

added test data for alteration scripts. should be good.

## Checklist

- [ ] unit tests
- [x] integration tests
- [x] necessary TSDoc comments

### CODE REVIEW

File: packages/schemas/src/types/tenant-organization.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    name: `Tenant ${tenantId}`,
+  });
+
+/**
+ * Scope names in organization template for managing tenants.
+ *
+ * @remarks
+ * Should sync JSDoc descriptions with {@link tenantScopeDescriptions}.
+ */
+export enum TenantScope {
```
Comment (gao-sun) [inline packages/schemas/src/types/tenant-organization.ts:51]: not finalized yet, but should not be a blocker of this pr

**Comment by wangsijie** [comment]
Want to make sure I understand correctly: in the previous(current online) design, users send requests (like `GET /applications`) directly to the tenant which they owned, and now, they can no longer touch the tenant API, instead, all requests go to the Cloud proxy.

**Comment by gao-sun** [comment]
> Want to make sure I understand correctly: in the previous(current online) design, users send requests (like `GET /applications`) directly to the tenant which they owned, and now, they can no longer touch the tenant API, instead, all requests go to the Cloud proxy.

@wangsijie correct

**Review by wangsijie — APPROVED** [review]

**Review by simeng-li — APPROVED** [review]

File: packages/schemas/alterations/next-1702544178-sync-tenant-orgs.ts [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    await transaction.query(sql`
+      insert into public.organization_role_scope_relations (tenant_id, organization_role_id, organization_scope_id)
+      values
+        (${adminTenantId}, 'owner', 'read-data'),
+        (${adminTenantId}, 'owner', 'write-data'),
+        (${adminTenantId}, 'owner', 'delete-data'),
+        (${adminTenantId}, 'owner', 'invite-member'),
+        (${adminTenantId}, 'owner', 'remove-member'),
+        (${adminTenantId}, 'owner', 'update-member-role'),
+        (${adminTenantId}, 'owner', 'manage-tenant'),
```
Comment (charIeszhao) [inline packages/schemas/alterations/next-1702544178-sync-tenant-orgs.ts:57]: According to the discussion yesterday, it seems we don't need to provide `owner` role? Only `admin` and `member` are predefined

Comment (gao-sun) [reply packages/schemas/alterations/next-1702544178-sync-tenant-orgs.ts:57]: will be a separate pr

**Pull request merged**